enforcer/utils/packetgen/packet_gen.go

```diff
@@ -0,0 +1,370 @@
+func (p *Packet) AddTCPLayer(srcPort layers.TCPPort, dstPort layers.TCPPort) error {
+
+	if p.TCPLayer != nil {
+		return fmt.Errorf("TCP Layer already exists")
+	}
+
+	//TCP packet header
+	p.TCPLayer = &layers.TCP{
+		SrcPort: srcPort,
+		DstPort: dstPort,
+		Window:  1505,
+		Urgent:  0,
+		Seq:     11050,
+		Ack:     0,
+		ACK:     false,
+		SYN:     false,
+		FIN:     false,
+		RST:     false,
+		URG:     false,
+		ECE:     false,
+		CWR:     false,
+		NS:      false,
+		PSH:     false,
+	}
+
+	p.TCPLayer.SetNetworkLayerForChecksum(p.IPLayer)
+
+	return nil
+
+}
+
+//SetTCPSequenceNumber changes TCP sequence number
+func (p *Packet) SetTCPSequenceNumber(seqNum uint32) error {
+
+	if p.TCPLayer.Seq != 0 {
+		return fmt.Errorf("Sequence number already exists")
+	}
+
+	p.TCPLayer.Seq = seqNum
+
+	return nil
+
+}
+
+//SetTCPAcknowledgementNumber changes TCP Acknowledgement number
+func (p *Packet) SetTCPAcknowledgementNumber(ackNum uint32) error {
+
+	if p.TCPLayer.Ack != 0 {
+		return fmt.Errorf("Acknowledgement number already exists")
+	}
+
+	p.TCPLayer.Ack = ackNum
+
+	return nil
+
+}
+
+//SetTCPWindow changes the TCP window
+func (p *Packet) SetTCPWindow(window uint16) error {
+
+	if p.TCPLayer.Window != 0 {
+		return fmt.Errorf("Window already exists")
+	}
+
+	p.TCPLayer.Window = window
+
+	return nil
+
+}
+
+//SetTCPSyn changes the TCP SYN flag to true
+func (p *Packet) SetTCPSyn() {
+
+	p.TCPLayer.SYN = true
+	p.TCPLayer.ACK = false
+	p.TCPLayer.FIN = false
+
+}
+
+//SetTCPSynAck changes the TCP SYN and ACK flag to true
+func (p *Packet) SetTCPSynAck() {
+
+	p.TCPLayer.SYN = true
+	p.TCPLayer.ACK = true
+	p.TCPLayer.FIN = false
+
+}
+
+//SetTCPAck changes the TCP ACK flag to true
+func (p *Packet) SetTCPAck() {
+
+	p.TCPLayer.SYN = false
+	p.TCPLayer.ACK = true
+	p.TCPLayer.FIN = false
+
+}
+
+//GetTCPSyn returns TCP SYN flag
+func (p *Packet) GetTCPSyn() bool {
+
+	return p.TCPLayer.SYN
+
+}
+
+//GetTCPAck returns TCP ACK flag
+func (p *Packet) GetTCPAck() bool {
+
+	return p.TCPLayer.ACK
+
+}
+
+//GetTCPFin returns TCP FIN flag
+func (p *Packet) GetTCPFin() bool {
+
+	return p.TCPLayer.FIN
+
+}
+
+//NewTCPPayload adds new payload to TCP layer
+func (p *Packet) NewTCPPayload(newPayload string) error {
+
+	if p.TCPLayer.Payload != nil {
+		return fmt.Errorf("Payload already exists")
+	}
+
+	p.TCPLayer.Payload = []byte(newPayload)
+
+	return nil
+
+}
+
+//GetTCPChecksum returns TCP checksum
+func (p *Packet) GetTCPChecksum() uint16 {
+
+	return p.TCPLayer.Checksum
+
+}
+
+//GetIPPacket returns IP checksum
+func (p *Packet) GetIPPacket() layers.IPv4 {
+
+	return *p.IPLayer
+
+}
+
+//GetTCPPacket returns created TCP packet
+func (p *Packet) GetTCPPacket() layers.TCP {
+
+	return *p.TCPLayer
+
+}
+
+//ToBytes creates a packet buffer and converts it into a complete packet with ethernet, IP and TCP (with options)
+func (p *Packet) ToBytes() [][]byte {
+
+	//Creating a ethernet packet with fixed MAC addresses
+	ethernetLayer := layers.Ethernet{
```

Comment:
This (EtheretHeader) should exist in the packet and allowed to be manipulated. Not here.